libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/map/MapViewBinder.kt

```diff
@@ -21,47 +18,57 @@ import com.mapbox.navigation.dropin.map.longpress.RoutePreviewLongPressMapCompon
 import com.mapbox.navigation.dropin.map.marker.MapMarkersComponent
 import com.mapbox.navigation.dropin.map.scalebar.ScalebarComponent
 import com.mapbox.navigation.dropin.navigationview.NavigationViewContext
-import com.mapbox.navigation.ui.app.internal.Store
 import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
-import com.mapbox.navigation.ui.app.internal.routefetch.RoutePreviewAction
-import com.mapbox.navigation.ui.app.internal.routefetch.RoutePreviewState
-import com.mapbox.navigation.ui.app.internal.routefetch.RoutesAction
 import com.mapbox.navigation.ui.base.lifecycle.UIBinder
 import com.mapbox.navigation.ui.maps.internal.ui.LocationComponent
 import com.mapbox.navigation.ui.maps.internal.ui.LocationPuckComponent
 import com.mapbox.navigation.ui.maps.internal.ui.RouteArrowComponent
 import com.mapbox.navigation.ui.maps.internal.ui.RouteLineComponent
-import com.mapbox.navigation.ui.maps.internal.ui.RouteLineComponentContract
 import com.mapbox.navigation.ui.maps.route.arrow.model.RouteArrowOptions
 import com.mapbox.navigation.ui.maps.route.line.model.MapboxRouteLineOptions
-import kotlinx.coroutines.FlowPreview
-import kotlinx.coroutines.flow.Flow
-import kotlinx.coroutines.flow.combine
 
 @ExperimentalPreviewMapboxNavigationAPI
-@FlowPreview
-internal class MapBinder(
-    private val context: NavigationViewContext,
-    private val binding: MapboxNavigationViewLayoutBinding,
-    private val mapView: MapView
-) : UIBinder {
-
-    init {
-        mapView.compass.enabled = false
+abstract class MapViewBinder : UIBinder {
+
+    companion object {
+        /**
+         * Default Info Panel Binder.
+         */
+        fun defaultBinder(): MapViewBinder = MapboxMapViewBinder()
+    }
+
+    internal lateinit var context: NavigationViewContext
+    internal lateinit var navigationViewBinding: MapboxNavigationViewLayoutBinding
+
+    private val mapboxMapObservers = mutableSetOf<MapboxMapObserver>()
+
+    internal fun registerMapboxMapObserver(observer: MapboxMapObserver) {
+        mapboxMapObservers.add(observer)
     }
 
-    private val scalebarComponent = ScalebarComponent(
-        mapView,
-        context.styles.mapScalebarParams,
-        context.systemBarsInsets
-    )
+    abstract fun getMapView(viewGroup: ViewGroup): MapView
```

Comment:
This method is public. Let's add KDoc for it.

What do you think about renaming this to `onCreateMapView(viewGroup: ViewGroup): MapView` so that it matches android framework (`Fragment.onCreateView`, `RecyclerView.onCreateViewHolder` etc.) and `InfoPanelBinder.onCreateLayout()`?

Reply:
Not sure about that. The MapView is added to layout in `addMapViewToLayout`. And `InfoPanelBinder.onCreateLayout()`, for example, is needed for creating and attaching mapView. So it would be easy to think that you should attach MapView in `onCreateMapView`.
Actually, I think that we can do it. But the customer would have to invoke
```
viewGroup.removeAllViews()
viewGroup.addView(mapView)
```
while before this PR we were doing it. But it seems reasonable.
What do you think?

Reply:
If we update `MapboxMapBinder` to either programmatically create the `MapView` or to use `LayoutInflater` instead of the Scene transition API, that would work. However, it looks a bit odd to me that `MapViewBinder` adds `MapView` differently than `MapboxMapBinder`. One might think that we made a mistake in `MapViewBinder.addMapViewToLayout ` and attempted to correct it in `MapboxMapBinder.addMapViewToLayout`.

Reply:
> However, it looks a bit odd to me that MapViewBinder adds MapView differently than MapboxMapBinder.

Agreed. But I don't fully understand what you are suggesting. Creating MapView from code looks a bit complicated. I'd leave inflating it from the xml.

Reply:
> > However, it looks a bit odd to me that MapViewBinder adds MapView differently than MapboxMapBinder.
> 
> Agreed. But I don't fully understand what you are suggesting. Creating MapView from code looks a bit complicated. I'd leave inflating it from the xml.

Before we had something like this:

```kotlin
abstract class MapViewBinder : UIBinder {
    ...
    abstract fun getMapView(viewGroup: ViewGroup): MapView

    internal open fun addMapViewToLayout(mapView: MapView, viewGroup: ViewGroup) {
        viewGroup.removeAllViews()
        viewGroup.addView(mapView)
    }
...
}

internal class MapboxMapViewBinder : MapViewBinder() {

    override fun getMapView(viewGroup: ViewGroup): MapView {
        Scene.getSceneForLayout(
            viewGroup,
            R.layout.mapbox_mapview_layout,
            viewGroup.context,
        ).enter()
        return MapboxMapviewLayoutBinding.bind(viewGroup).mapView
    }

    override fun addMapViewToLayout(mapView: MapView, viewGroup: ViewGroup) {
        // do nothing
    }
    ...
}
```

It looked odd to me that we implemented `addMapViewToLayout()` method in `MapViewBinder` and did not use it in a subclass `MapboxMapViewBinder`. I see that you already updated it and moved the responsibility of adding MapView to the ViewGroup to `onCreateMapView()` implementation. I am fine with that as long as we mention it in method KDoc.

What I was suggesting was something like this:

```kotlin
abstract class MapViewBinder : UIBinder {
    ...
    abstract fun getMapView(viewGroup: ViewGroup): MapView
     
    @MapStyleLoad.Policy
    open fun getMapStyleLoadPolicy(): Int = MapStyleLoad.POLICY_NEVER
   
    override fun bind(viewGroup: ViewGroup): MapboxNavigationObserver {
        val mapView = onCreateMapView(viewGroup)
        viewGroup.removeAllViews()
        viewGroup.addView(mapView, MarginLayoutParams.MATCH_PARENT, MarginLayoutParams.MATCH_PARENT)
        ...
    }
}

@ExperimentalPreviewMapboxNavigationAPI
internal class MapboxMapViewBinder : MapViewBinder() {

    override fun onCreateMapView(viewGroup: ViewGroup): MapView =
        MapView(viewGroup.context).apply {
            compass.enabled = false
        }

    @MapStyleLoad.Policy
    override fun getMapStyleLoadPolicy(): Int = MapStyleLoad.POLICY_ON_CONFIGURATION_CHANGE
}
```

Reply:
Thanks for the detailed response. I like your idea with a few notes:
1) Is ```
MapView(viewGroup.context).apply {
    compass.enabled = false
}
```
everything that's needed to create MapView? Because in `customMapViewFromCode` in qa-test-app it's much more complicated.
2) If we add the MapView to layout ourselves (which I like: if we can make user's life easier, we should), I guess it shouldn't be named `onCreateMapView` anymore (because in the analogous `InfoPanelBinder#onCreateLayout` it is required that it's inflated by the user while we only need a reference to MapView so it may be confusing). What if I change it back to `getMapView`?

Reply:
1. Yes.
Take a look at the XML layout. Our default MapView relies on default `MapInitOptions`.
```
<?xml version="1.0" encoding="utf-8"?>
<merge
    xmlns:android="http://schemas.android.com/apk/res/android"
    xmlns:tools="http://schemas.android.com/tools"
    tools:parentTag="android.widget.FrameLayout"
    >
    <com.mapbox.maps.MapView
        android:id="@+id/mapView"
        android:layout_width="match_parent"
        android:layout_height="match_parent" />
</merge>
```
2. Fine by me.